### PR TITLE
Improve GPT3.5 reasoning and adds function calling agent

### PR DESCRIPTION
**Pull request by nadheesh**
## Purpose
resolves: https://github.com/wso2-enterprise/choreo/issues/25849, https://github.com/wso2-enterprise/choreo/issues/23377

### CODE REVIEW

File: ballerina/agent.bal [hunk 1]
```diff
@@ -84,271 +125,162 @@ public class AgentExecutor {
...
+    # + return - Llm tool response, chat response or an error if reasoning has failed
+    public isolated function reason() returns LlmToolResponse|LlmChatResponse|TaskCompletedError|LlmError {
         if self.isCompleted {
-            return error TaskTerminationError("Task is already completed. No more reasoning is needed.");
+            return error TaskCompletedError("Task is already completed. No more reasoning is needed.");
         }
-        string|error decision = self.decideNextTool();
-        if decision is error {
-            return error LlmGenerationError("Error while communicating to LLM.", decision);
+        LlmToolResponse|LlmChatResponse respond = check self.agent.selectNextTool(self.progress);
```
Comment (SasinduDilshara) [inline ballerina/agent.bal:135]:
```suggestion
        LlmToolResponse|LlmChatResponse response = check self.agent.selectNextTool(self.progress);
```
wdyt?

File: ballerina/agent.bal [hunk 2]
```diff
@@ -84,271 +125,162 @@ public class AgentExecutor {
...
+                observation = output;
+            }
         }
-        any|error observation = self.toolStore.runTool(nextTool.tool, nextTool.tool_input);
-        if observation is ToolNotFoundError || observation is ToolInvalidInputError {
-            map<value:Cloneable> detail = observation.detail();
-            observation = detail.hasKey(ERROR_INSTRUCTION_KEY) ? detail.get(ERROR_INSTRUCTION_KEY) : "Tool execution failed due to invalid generation. Regenerate following the given format.";
-        } else if observation is () { // whether tool return any value or error 
-            observation = string `Tool '${nextTool.tool}' didn't return any values or errors. Verify whether it was successful.`;
+        else {
```
Comment (SasinduDilshara) [inline ballerina/agent.bal:163]: this else should be

```
} else {
```

File: ballerina/agent.bal [hunk 3]
```diff
@@ -84,271 +125,162 @@ public class AgentExecutor {
...
-Question: [The input question you must answer]
-Thought: [You should always think about what to do]
-Action: [Select a single tool from the provided list and use the following format within backticks. This field is mandatory after 'Thought'.]
-${BACKTICK}${BACKTICK}${BACKTICK}
+        iter += 1;
+        if verbose {
+            io:println(string `${"\n\n"}Agent Iteration ${iter.toString()}`);
+            SelectedTool|LlmInvalidGenerationError tool = step.toolResponse.tool;
+            if tool is SelectedTool {
+
```
Comment (SasinduDilshara) [inline ballerina/agent.bal:239]:
```suggestion
```

File: ballerina/constants.bal [hunk 4]
```diff
@@ -27,7 +28,14 @@ const OPENAPI_PATTERN_DATE = "yyyy-MM-dd";
 const OPENAPI_PATTERN_DATE_TIME = "yyyy-MM-dd'T'HH:mm:ssZ";
 
 //agent
-const FINAL_ANSWER_KEY = "final answer";
+// const FINAL_ANSWER = "final answer";
```
Comment (SasinduDilshara) [inline ballerina/constants.bal:31]: we can remove this comment I guess

File: ballerina/llm.bal [hunk 5]
```diff
@@ -246,44 +270,37 @@ public isolated class AzureChatGptModel {
...
+            stop,
+            messages,
+            functions: from AgentTool tool in tools
+                select {
+                    name: tool.name,
+                    description: tool.description,
+                    parameters: tool.variables
+                }
+        };
+        azure_chat:CreateChatCompletionResponse response = check self.llmClient->/deployments/[self.deploymentId]/chat/completions.post(self.apiVersion, request);
```
Comment (SasinduDilshara) [inline ballerina/llm.bal:291]: Character count exceeds 120

File: ballerina/react.bal [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            return llmResult;
+        }
+        return error LlmConnectionError("Geneartion Failed.", llmResult);
+    }
+
+}
+
+isolated function normalizeLlmResponse(string llmResponse) returns string {
+    string normalizedResponse = llmResponse.trim();
+    if !normalizedResponse.includes("```") {
```
Comment (SasinduDilshara) [inline ballerina/react.bal:85]:
``` can be a constant.

File: ballerina/react.bal [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    string historyPrompt = "";
+    foreach ExecutionStep step in history {
+        string observationStr = getObservationString(step.observation);
+        string thoughtStr = step.toolResponse.llmResponse.toString();
+        historyPrompt += string `${thoughtStr}${"\n"}Observation: ${observationStr}${"\n"}`;
+    }
+    return historyPrompt;
+}
+
+isolated function constructReActPrompt(ToolInfo toolInfo) returns string {
```
Comment (SasinduDilshara) [inline ballerina/react.bal:150]: This function can be a arrow function

File: ballerina/tests/agent-test.bal [hunk 8]
```diff
@@ -37,51 +37,67 @@ function testAgentInitialization() {
...
 @test:Config {}
 function testAgentExecutorRun() returns error? {
-    Agent agent = check new (model, searchTool, calculatorTool);
+    ReActAgent agent = check new (model, searchTool, calculatorTool);
     string query = "Who is Leo DiCaprio's girlfriend? What is her current age raised to the 0.43 power?";
-    AgentExecutor agentExecutor = agent.getExecutor(query);
-
-    record {|ExecutionStep|error value;|}? result = agentExecutor.next();
+    AgentExecutor agentExecutor = new (agent, query);
+    // _ = run(agent, query);
```
Comment (SasinduDilshara) [inline ballerina/tests/agent-test.bal:95]: I think we can remove this comment

File: ballerina/tests/mock-tools.bal [hunk 9]
```diff
@@ -31,49 +38,62 @@ isolated function calculatorToolMock(*CalculatorParams params) returns string {
     }
 }
 
-public client class MockLLM {
+isolated function sendMail(record {|string senderEmail; MessageRequest messageRequest;|} 'input) returns string|error {
```
Comment (SasinduDilshara) [inline ballerina/tests/mock-tools.bal:41]: input is not a keyword in Ballerina

File: ballerina/constants.bal [hunk 10]
```diff
@@ -27,7 +28,14 @@ const OPENAPI_PATTERN_DATE = "yyyy-MM-dd";
...
 const THOUGHT_KEY = "Thought:";
-const BACKTICK = "`";
+const BACKTICKS = "```";
 const ERROR_INSTRUCTION_KEY = "instruction";
+
+final regexp:RegExp FINAL_ANSWER_REGEX = re `^final.?answer`;
```
Comment (LakshanWeerasinghe) [inline ballerina/constants.bal:36]:
```suggestion
final string:RegExp FINAL_ANSWER_REGEX = re `^final.?answer`;
```

File: ballerina/agent.bal (hunk 2 above)
Comment (nadheesh) [reply ballerina/agent.bal:163]: fixed, why formatter didn't catch it?

File: ballerina/constants.bal (hunk 4 above)
Comment (nadheesh) [reply ballerina/constants.bal:31]: fixed

File: ballerina/llm.bal (hunk 5 above)
Comment (nadheesh) [reply ballerina/llm.bal:291]: fixed

File: ballerina/react.bal (hunk 6 above)
Comment (nadheesh) [reply ballerina/react.bal:85]: resolved

File: ballerina/react.bal (hunk 7 above)
Comment (nadheesh) [reply ballerina/react.bal:150]: addressed

File: ballerina/tests/agent-test.bal (hunk 8 above)
Comment (nadheesh) [reply ballerina/tests/agent-test.bal:95]: fixed

File: ballerina/agent.bal (hunk 2 above)
Comment (SasinduDilshara) [reply ballerina/agent.bal]: I did some research about this, seems like both are ok in programming.

**Review by SasinduDilshara — APPROVED** [review]

File: ballerina/tests/mock-tools.bal (hunk 9 above)
Comment (nadheesh) [reply ballerina/tests/mock-tools.bal:41]: fixed

**Pull request merged**